### DISCUSSION TITLE
Variable size multidimensional arrays - shorter declaration

### DISCUSSION THREAD

**Discussion by mobarski** [vlang/v#11704] [Ideas]
Currently we can declare variable size multidim arrays like this:

```v
// 2d
mut a := [][]int{len:2, init: []int{len:3}}

// 3d
mut a := [][][]int{len:2, init: [][]int{len:3, init: []int{len:2}}}

// ...
```

Because arrays are frequently used we should have additional shorter syntax for declaring them (Alex M. agrees).

The exact syntax is open for discussion.

My proposal:
- basic use: `[][][]int{len:[2,3,2]}`
- advanced use `[][][]int{len:[2,3,2], init:1, cap:[10,20,10]}`

**Comment by ntrel** [comment]
I'm not sure this is common enough to have special syntax. We could have a generic function that does the same thing:
```v
a := md_array<[][]int>(len: [2, 3])
```
While that function can't be implemented ATM, V's compile-time abilities should be extended to support it.

Also, `init` should be an array too.

**Comment by changrui** [comment]
Hi, all, how to generate the following mutli-dimensional array after `mut arr := [][]int{}`, thanks so much.
```
[ [ 1 ]
  [ 1, 1 ]
  [ 1, 2, 1 ]
  [ 1, 3, 3, 1 ]
  [ 1, 4, 6, 4, 1 ]
  [ 1, 5, 10, 10, 5, 1 ]
]
```

**Comment by petaqb** [comment]
@changrui     Hi, you can do it like that:

```
module main

fn main() {
	mut table := [][]int{} // array of rows of different length
	mut row := []int{} // row is array of an any number of ints
	table << [1] // first member

	mut val := 0

	for m := 1; m < 6; m++ { // starting from second member
		row = []
		row << 1
		for n := 0; n < m - 1; n++ {
			val = table[m - 1][n] + table[m - 1][n + 1]
			row << val // initialize rows
		}

		row << 1
		table << row
	}

	dump(table)
	// gives table=
	[[1],
	[1, 1],
	[1, 2, 1],
	[1, 3, 3, 1],
	[1, 4, 6, 4, 1],
	[1, 5, 10, 10, 5, 1]]
}
```

**Comment by changrui** [comment]
@mobarski I'am so sorry for reply delaying. Thanks so much, I got it..

**Comment by petaqb** [comment]
Perhaps, for low dimensional array one could consider as more V-styled the following:
```v
mut a [][][]int{len:[2][3][2], init:1, cap:[10][20][10]}

mut a [][]int{len:[2][2], init:[[11,12],[21,22]]}
```
For larger dimensions one could use some fn() which will create an array in needed form, as in my example for Pascal's triangle
